### ISSUE TITLE
Router: Design for routerCanActivate, an injectable alternative to @CanActivate

### ISSUE THREAD

**Issue by wardbell** [angular/angular#7784]
This issue describes a design for an injectable alternative to `@CanActivate`

It is both a fresh start and a continuation of the conversations associated with issues #748 and #4112 and tangentially in #7256, #2965, #7091

---

I've been talking to the team about these router hook methods. I have a good feeling about an emerging design that should arrive before RC. I thought I'd share it here so we can get some feedback.

CAVEAT: this is *Ward's* interpretation of what he _thinks_ he heard which may vary from the team's interpretation and is not necessarily what will happen.

As I understand it, `CanActivate` will become an instance method, probably named **`routerCanActivate`**.

The catch is that we implement `routerCanActivate` on (what I call) a ***routing component***: a component with an attached router and routes and whose template includes a `<router-outlet>`.

>We didn't talk about it but I think Angular should ignore it on a non-routing component or perhaps throw an error.

The router will consult this method before pursuing any of the routes that would populate that `<router-outlet>`. It gates navigation to any of the routes configured for this component's router.

To be clear, it has no role in allowing/preventing navigation to _the routing component itself_; only navigation to the routes configured for this routing component's `<router-outlet>`.

As an instance method, the `routerCanActivate` has access to anything we inject into its routing component. The injector (and its scope) is that of this routing component.

I think this addresses many concerns:

1. We can inject anything we need without crazy hacks

2. We can scope injection at an appropriate level in the component tree. We don't have to capture the root injector and hope that the service we need has been provided there. We can access services injected "down here", at the local router configuration level.

3. We can make decisions about whether to allow navigation to a lazy loaded component. Before we would have had to load that component and call its `@CanActivate`, perhaps only to belatedly disallow it.

I had a few initial objections:

1. I didn't think the routing component should be responsible for *can activate* decisions about all of its routes.

2. Why should it know all of the details about whether a route's navigation should be allowed? That's a decision that belongs to the destination component, right?

3. Some people have used `CanActivate` to pre-load data from the server for the destination component.  There won't be an obvious way to do that with this design which doesn't provide a way to deliver a payload to the destination component.

4. The *can activate* logic has access to services at the routing component level and above but not the services (if any) that will be provided by (and therefore scoped at) the destination component itself. That could be too limiting.

Here is why these objections no longer trouble me.

1. If deciding is a burden, the routing component can delegate that burden to a custom injected `CanActivateService`, scoped to the routing component itself. That service can inject much of what it needs to make decisions. The routing component can stay lean. This is *our* decision as application developers; Angular 2 neither requires it nor blocks it.  I like that.

2. While I like the idea that the *can activate* logic is part of the development of the destination component, that logic doesn't have to be inscribed within the destination component class. We can come up with other patterns that will work well, particularly in concert with the `CanActivateService` approach I just mentioned. I think we can get the allocation of responsibilities right without too much friction. And this approach does put us in position to gate navigation to a lazy-loaded destination component which I think is the more important benefit.

3. Actually, I always thought using *can activate* logic to pre-load data was a hack. Many folks did that kind of thing with resolve in Angular 1.  I never liked it.  Please don't bring back `resolve`!

  But like/dislike aside, we don't need it in Angular 2.  In Angular 2, we have both the `routerOnActivate` and the `ngOnInit` hooks (missing in Angular 1) which are fine places to pre-load data.

4. I couldn't think of a use case in which it was essential that we consult a service instance created just for the destination component. OTOH, if we _could_ do that, we'd be faced with the obligation to create such a service which brings unwanted complexity of its own. 

  For example, if we allow navigation, we have to make sure that service (whatever it is), is ultimately injected into the destination component. There is no facility for doing that kind of thing today. Today we can't access the internal steps that go into hydrating a component ... and I rather doubt I'd want to risk what comes with breaking into that sausage factory. 

  What if we created that service and then _rejected_ the navigation? What did that service do? Did it know it was going to be created and then thrown away? Does it need to be cleaned up before being thrown away.  It's a horrible can of worms.

In the end, I think this design will work well for us. It's safe, clean, easy to understand once explained.

**Angular consults a routing component's  `routerCanActivate` method before navigating to any of the routes defined for that component's router.**

I don't know what the method parameters will be. Obviously something about where the router is going.  This is a good time for you all to chime in.

p.s.: I don't know what will become of the current `@CanActivate` decorator. I suppose it could live on as a vestigial relic. Personally, I'd vote for it to be deprecated immediately and removed before release.

p.p.s.:  I have stopped thinking that the `routerOnActivate` method is or could be a substitute for `routerCanActivate` (quite apart from the fact that it doesn't work as a substitute today).

  It should not be involved in the decision about whether we should navigate to this component. It should concentrate on what to do _now that we're here_. And in that respect it should differentiate from its companion hook method, `routerOnReuse`, which concerns what to do _now that we're back_.

**Comment by brandonroberts** [comment]
@wardbell I like this idea. I was just working through a plunker that does something along these lines. The `AuthRoute` in the plunker would be considered as your **routing component** and the `routerOnActivate` would be replaced with your `routerCanActivate`. Let me know if what I have is off base.

http://plnkr.co/edit/tUhNqM2cNDp20ch7cezH?p=preview

**Comment by wardbell** [comment]
@brandonroberts First glance (I didn't go deep), you're spot on. And you wouldn't have to "abuse" `routerOnActivate` in this fashion. It could truly be what it is supposed to be ... an asynchronous method for handling component activation.

Aside: I have my quarrel with the difference between the way the promise is treated in a *routing component* vs. a *leaf component*, differences I explore in a plunker and commentary in #7576. I am hopeful that we can clear that up too!

I'm starting to feel good about how the router will land! :-)

**Comment by petebacondarwin** [comment]
@wardbell Angular 1 does now have `$onInit` along with `$onDestroy`, `$postLink` and `$onChanges` hooks for component directives now :-)

**Comment by choeller** [comment]
@wardbell I really like this idea! It would totally solve all problems currently present with the Hook being configured at the target-component. Furthermore I think it would be super useful to have something like the `routerOnActive`-Hook on the **routing component** level, as currently (AFAIK) there is no clean way to react on the navigation to another router. 
For example when trying to implement a generic method to set the browser-title based on the RouteData configured for a Route I currently need to use the `router.subscribe`-method and then get the current instruction via "router.currentInstruction". This is super hacky, as you need to deal with child-components etc. yourself (if interested see my approach in the comment in this issue: https://github.com/angular/angular/issues/7630) 
Having the possibilty to just implement a method telling me - "I'm now navigating from comp1 to comp2" would be really neat!

**Comment by CaptainCodeman** [comment]
My possibly simplistic, likely ignorant but still opinionated view of routing:

We already have a method for injecting dependencies - the constructor.
We already have a method for setting input values - the @Input attribute.
We already have a method for setting output values - the @Output attribute.

Why can't the routing just use what is already there? Why does a component need to know anything at all about routing (other than when it wants to say "go to this route")? How it's hosted and where it's parameters came from (the URL or another intermediate component) shouldn't matter. Why can't it output something to say failed or not-allowed that the route-outlet or other hosting component can use to redirect elsewhere if required? I just want the router to be the mapping between the URL and my components inputs.

CanReuse can be decided by how the component responds to inputs changing (no different to any other 'Smart Component'). CanActivate just seems to be OnInit with another name. All the different overlapping component / router lifecycle methods just add to confusion and mean refactoring if you move around how your components are used.

How "resolve" happens is really the job of the components - one smart component wrapping another if needed.

**Comment by wardbell** [comment]
@choeller Thanks for those observations. We have feature issues for both the concerns that you raise: (1) how to set the browser title bar and the back-button titles (!) and (2) how to listen to a router in order to respond when it does something.  Having to manually parse the url from `router.subscribe` is pretty awful and I'm glad you're drawing more attention to that.

@CaptainCodeman  There are no "simplistic" or "stupid" questions. If you need to ask it, someone is grateful you did.

I'll start with your second question: "*Why does a component need to know anything at all about routing ...?*"

We have to write the logic that decides whether (and how) a navigation request should proceed. I think you're asking about _where_ we put that logic: should we put it in the routing component or attach it to the router?

We could ask the same question about route configuration: should we put that configuration in the routing component or attach it to the router directly?

My instinctive answer ... much like yours ... is to put it on the router.  I don't have a problem attaching a router to a component later (thus making it a "routing component"). But it does feel like a separate concern.

I don't have a great answer. The public router API suggests that you can create a router, configure it and attach it to a component. But no one demonstrates that technique. Why?

I vaguely recall this being discussed and I think the answer was something like "*you can't use a router independently of a component so let's make it easy and do both router configuration and component definition in the same place.*"  That makes some sense. To be honest, I stopped thinking about it because that ship sailed a long time ago.

You have a nest of other questions that surround this one. I'm not clear about what you're asking.

*  _Why can't the routing just use what is already there?_   Use how? The proposal says that the `routerCanActivate` implementation will have access to the input, output and injected material. What is unclear?

* _How it's hosted and where its parameters came from (the URL or another intermediate component) shouldn't matter._  Shouldn't matter to whom? It matters to the router. It matters if the component must update a visual element based on the current state of the router.

* _Why can't it output something to say failed or not-allowed ...?_  What is the "it" in that question. Output to what? What would the API look like? Sorry but I don't know what you're asking.

* _I just want the router to be the mapping between the URL and my components inputs._ What do you mean by a mapping? How should a URL map to a component's input properties? What difference would that make? And is that _really_ all you want from the router? It seems to me - based on my experience and many other discussions - that we want it to do a lot more, including maintain the browser window stack, handle parameters, wait for other actions to complete, add classes to elements, update title bars ... etc.

I think I understand better what you say after that paragraph.

_CanReuse can be decided by how the component responds to inputs changing (no different to any other 'Smart Component')._

Let's look at why `CanUse` exists. The router sometimes can re-use a component instance when it navigates through that component type again.  The rules are a bit murky to me. Apparently there are some benefits to such reuse.  But it isn't always a good idea to re-use that instance even if the router is allowed to do so. We can prevent the router from trying to re-use it by setting `CanReuse === false`. That's a necessary option.

If the router reuses a component instance, it calls `routerOnReuse` so that the component can adjust accordingly. If the router creates a new instance _instead_, it calls `routerOnActivate` which is likely to require quite different logic. I see value in that distinction. I hope you do.

 _CanActivate just seems to be OnInit with another name._   This statement confounds me. `CanActivate` and the future `routerCanActivate` make decisions that influence router behavior: can we navigate to the destination component and when.  Whereas `ngOnInit` performs initialization of a newly created component; it doesn't and can't make navigation decisions.  And we need `ngOnInit` for components that don't participate in routing at all.

I'm not sure if you really meant `CanActivate`.  Maybe you're saying that `routerOnActivate` and `ngOnInit` seem the same. Indeed, they both are called when the router creates a new instance. For me, the critical difference is that `routerOnActivate` supplies information about the navigation that may be important to component initialization. The route parameters are an important example: I can't initialize a component that will display a `Hero` without knowing which hero.  Other aspects of the `routerOnActivate` parameters may be important to the initialization process as well. This routing info isn't available through input properties so there is no way for `ngOnInit` to know about them.

>Yes, we can inject `routeParams` separately. But that seems like clutter and we wouldn't have access to the other routing info that are only available through the `routerOnActivate` parameters.

_All the different overlapping component / router lifecycle methods just add to confusion and mean refactoring if you move around how your components are used._  They sure are confusing ... and will remain so until we explain what they do, when they do it, and why.  That's a documentation effort that awaits the router repairs currently underway. We will get there.

_How "resolve" happens is really the job of the components - one smart component wrapping another if needed._  Absolutely agree.

**Comment by CaptainCodeman** [comment]
I guess the premise of the questions / points I was making is: why does a routed component suddenly have to switch to using completely different mechanisms for all manner of things that already exist and work perfectly well. Components already need to know when they are initialized, when their inputs change and to get access to those input parameters - why should any of them need to change just because routing is involved?

> I vaguely recall this being discussed and I think the answer was something like "you can't use a router independently of a component so let's make it easy and do both router configuration and component definition in the same place." That makes some sense. To be honest, I stopped thinking about it because that ship sailed a long time ago.

I think there is a significant difference between a component that is being routed *to* and a component that is itself using routing and acting as a parent to others. The latter absolutely needs to know about routing and configure both the routes and the outlet for it to use (which the current route config does well). That makes perfect sense, but I don't understand why the child component that will be _loaded_ needs to have any references to the router and need to provide new and different routing specific methods for it. Instead I'd like the router to use the mechanisms that the rest of the framework and other components use.

As a simplistic example, suppose we have a component that accepts an id and renders some UI for it. It's small so is inline:

```
<li *ngFor="#item of items">
  <an-item [id]="item.id"></an-item>
</li>'
```

Later on it grows and we decide to change things so that the items are rendered as a list with links to view the larger component either on a separate page or rendered at the side in a router outlet. We add some route config and change the list to use routerLink:

```
<li *ngFor="#item of items">
  <a [routerLink]="['Item', { id: item.id }]">{{ item.name }}</a>
</li>'
```

We can't currently simply reuse the component that is already working fine. Instead we have to re-write it because the router can't give it the id it was expecting using the existing `@Input` mechanism. We have to start adding other methods to optimize re-use to stop the screen flashing when each item is clicked and to also pick out the input parameter from the routeParams or the instruction (differently for each). Incidentally, this named parameter is also no longer defined by the component, it's defined externally in the routing which IMO makes the component slightly less reusable as well as being coupled to the router and reverses the normal direction. If the component also uses services passed into the constructor then we're now talking about a different mechanism for injecting those too?

All these changes work both ways - if we decide to move a previously rooted component inline then it all changes the other way.

We could create our own wrapper to avoid this - accept the router values and host the child component and pass the values as inputs. But why can't the router just do this directly? It's the sort of thing that just adds additional complexity and confusion or some frustration with using the framework which I think could possibly be avoided.

> Why can't the routing just use what is already there? Use how? The proposal says that the routerCanActivate implementation will have access to the input, output and injected material. What is unclear?

I'm not saying I don't understand how to use it. I'm saying I don't understand why it needs to be this way. I think the discussion about how to create a new DI mechanism is only happening because of how the router has been implemented to where it is now.

Taking each piece separately:

Injecting dependencies into a component: use the constructor. I don't know why I'd ever want or need another way to do it as well and what this would add other than confusion.

Passing parameters into a component: `@Input()`. This also handles the parameters changing and re-use in that whatever the component wants to do it can do with no need to tell the router that it wants to be re-used (that should be in the routing config IMO).

The only output seems to be whether the component can be activated or not. This could be an `@Output` or some return value from the input changes method.

> I just want the router to be the mapping between the URL and my components inputs. What do you mean by a mapping? How should a URL map to a component's input properties? What difference would that make? And is that really all you want from the router? It seems to me - based on my experience and many other discussions - that we want it to do a lot more, including maintain the browser window stack, handle parameters, wait for other actions to complete, add classes to elements, update title bars ... etc.

If I have a route defined in a parent that has chosen to use routing and it has a url such as `/mythings/:id` then whatever is in `:id` gets passed into the component's `@Input() id:string` property.

Of course I want the router to maintain history and all that "routery stuff" - but what has any of that got to do with my component? All *it* wants to know about is that it needs to do it's stuff and has been given the id that it needs.

re: the whole CanActivate thing - I'll reply to that separately when I have more time.

**Comment by petebacondarwin** [comment]
@CaptainCodeman - it is interesting reading your ideas here as they sound very similar to the motivation behind the new Polymer Router: https://www.polymer-project.org/1.0/articles/routing.html

**Comment by CaptainCodeman** [comment]
I heard "you may not actually be insane". Wohoo ! :grinning: 

Thanks, I'll give that a read. It's a while since I last looked at polymer.

**Comment by ericmartinezr** [comment]
@wardbell regarding about your answer to @CaptainCodeman 

>The proposal says that the routerCanActivate implementation will have access to the input, output and injected material. What is unclear?

I can't see how that would work in the code. If you have a parent component and your children emit some event (@Output & EventEmitter) how the parent will be react to it? We don't add the routed components in the view. The same for the @Input. Or I misunderstood that part?

I want to say that the whole proposal looks nice from the point of view from someone who started all this web development with ng2 and has barely a month of playing with ng1 and never used router with it.

I just would love (I hope I understood it fine, sorry if I didn't, didn't want to wait to all this is released and they say : *"Why in the hell I didn't say nothing..."*) routed components behave like normal components. Communicating routable components with parent components through services is tooo much work.

**Comment by robwormald** [comment]
I've never particularly liked the coupling between a component and the router. I'd prefer to have these sorts of things in the RouteConfig:

```typescript
@RouteConfig([
  {
    name: 'AuthedRoute',
    path: '/auth',
    component: AuthComponent,
    canActivate(){  return Promise.resolve('ok') } 

])
```

For things that require services:

```typescript
@RouteConfig([
  {
    name: 'AuthedRoute',
    path: '/auth',
    component: AuthComponent,
    canActivate: [(authService, routeParams) => authService.check(routeParams.get('id')), AuthService, RouteParams]

])
```

The actual activation happens scoped to the router outlet, which appears to have access to the appropriate injector. Just a thought, seems to make things more reusable.

**Comment by CaptainCodeman** [comment]
> Let's look at why CanUse exists. The router sometimes can re-use a component instance when it navigates through that component type again. The rules are a bit murky to me. Apparently there are some benefits to such reuse. But it isn't always a good idea to re-use that instance even if the router is allowed to do so. We can prevent the router from trying to re-use it by setting CanReuse === false. That's a necessary option.

That describes the 'what' but not really the 'why'. I've found I use CanReuse = true for everything because it seems to be asking if I want the screen to flicker or not. I haven't come across a case where I ever wanted it to be false.

I see it as either a cheap "reset everything for me" or else it's todo with caching and memory control. I hate that the default is the "slow" version and also having to add it to cater for what I imagine must be components that don't depend on the inputs (not sure how that even works) and if it's the latter then I'd rather be able to configure something on the  router to say "keep x components cached" or base it on memory use automagically.

Fundamentally, I think we need more concrete examples of usage than "someone may want to use it" because that would justify any option on any API.

The only piece that I think doesn't have a direct and immediate existing in-component counterpart is the CanActivate piece. I don't use CanActivate but do have my route permissions defined on the route config (picked up and handled by a secure router outlet) which is closer to @robwormald example above. These are just simple public flag + allowed role checks though (against JWT token roles) and any auth services it uses are injected into the outlet.

I'm not convinced that having the whole CanActivate and injection mechanism + code on the router config is worth the complexity though. If the typical use case is to simply have some per-object permission check then I'd prefer to do that in my component with the existing injected services and input parameters going through the existing code-paths and then raise some event to signal 'rejection' instead of rendering than to possibly duplicate pieces of it elsewhere.

Could the route config be told to check for an output event or return something from the onChanges call to indicate rejection?

I'm actually unclear what the router currently does if CanActivate fails - are we supposed to define an alternate route for it to go to instead? One of the properties I have on my router outlet is the route to go to if permission is denied - where is that in the normal router config / CanActivate system?

Does it just not render anything? For me that wouldn't be enough - it needs to render something or redirect someplace to give the necessary UI feedback.

**Comment by laurelnaiad** [comment]
@CaptainCodeman I believe it was stated that CanActivate is determined while selecting the best matching route, so the process would just move on if "Can't" Activate is true for any given route alternative.

I think many of us will want CanReuse=true to be our default answer, and I hope we don't need to write a function to say so... so that if CanReuse isn't the default (I kinda hope it is the default), then at least it could be settable by a simple property in the decorator.

**Comment by brandonroberts** [comment]
@CaptainCodeman @laurelnaiad My use case for returning false from CanReuse, was to force a reload of an entire route from top to bottom as discussed here: https://github.com/angular/angular/issues/7429

As far as routing components being treated as regular components, I think the issue goes a little deeper than the router and goes back to the dynamic component loader, which is what the router uses to load components for a route. If I'm not mistaken its a design decision not to allow dynamically loaded components to not react to inputs or emit outputs as it adds too much unpredictable behavior when inserting a component into the view.

As far as the proposal goes, I think the router API also needs to allow for more granular control over which routes you want to protect or delay from activation. @choeller already referred to it, but I personally don't want to setup a complicated tree of routes for the sake of keeping some routes public and some routes private. The `routerCanActivate` needs to be more aware of where you are going as well as where you are. This has seem to go against the current design of the router, which is to keep route components ignorant to their surroundings during navigation, including what their parent is, what their children are and where you are currently during the navigation lifecycle.

**Comment by CaptainCodeman** [comment]
Thanks @laurelnaiad I guess the drawback of that is that you end up duplicating route configs (or can you just have a single catchall redirect to a "permission denied" component? How does that work with "not found"? I don't want to have two route config entries for every route - one to say "can I?" and the other for when the answer is no.

@brandonroberts Having the default behaviour for CanReuse dictated by (IMO) a somewhat dubious need to reload the entire route seems very, very wrong to me - surely the point of Angular and a SPA is that this isn't necessary, the components should just update the pieces that are needed as state changes. If a change at the top-level ends up causing a cascade of other changes then fine - but I don't understand what the router is doing being involved in that or how CanReuse really helps with any of it (heck, why not just trigger a page refresh?). At the very least it should be defaulted to true so that it doesn't have to be overridden for the 99% of the time that it isn't needed (yeah, I possibly made that stat up after a poll of one person).

I don't understand enough of the internal workings to understand why a component couldn't be loaded and the `@Input()` parameters used instead of the same component loaded and the same parameters passed into it via a method instead.

**Comment by brandonroberts** [comment]
@CaptainCodeman the idea is that you need to reload parent route components based on some action that happens in a child route component. Causing a full reload would not be a desired option in my opinion as it is an SPA.

Maybe I'm missing something about how you want the component reuse to be handled. You want the route the be reused by default no matter what when navigating to the same route multiple times (e.g. pagination)?

**Comment by CaptainCodeman** [comment]
> the idea is that you need to reload parent route components based on some action that happens in a child route component. Causing a full reload would not be a desired option in my opinion as it is an SPA.

Exactly, that's the point I was exaggerating to make - a full page reload is totally wrong but so is forcing a heavy-handed reload of anything else in a SPA including parent routes. It should be controlled based on the inputs to a component or some other state it watches, not because some child happens to return false to a call to CanReuse - no wonder it made the page flicker if that's what the default does.

In terms of re-use I would prefer the default behaviour to assume that the app is doing things properly and not needing blanket refreshes to happen. Everything should be based on inputs and state - if things rely on pieces that don't change then nothing happens, if they rely on inputs that have changed then those pieces will update. Isn't that the whole basic premise of Angular 2 components?

Pagination is actually a good example. While implementing pagination + filtering it felt like I was fighting with the router as much as I felt it was helping me. What should be simple - updating state and making the URL change to reflect it without things reloading seemed more difficult than it should be. Maybe I'm just using it wrong but I spent quite a bit of time trying different approaches and compared to the rest of Angular 2 it just never felt elegant or intuitive.

**Comment by wardbell** [comment]
@CaptainCodeman I'd like to dig into everything you have said. At the moment, I only have the strength to comment on `routerCanReuse`.

In my experiments, setting it to `true` is a complete waste of time because:
1. it is already true by default;
2. the router only re-uses an instance under certain specific circumstances; 
3. setting `routerCanReuse` to true won't force the router to re-use an instance if it thinks it can't.

That's why I say the only thing you can really do is set it false when you want to ensure that the router _does not reuse an instance_.

Do you have any evidence to the contrary? Show me the code please.

**Comment by choeller** [comment]
@CaptainCodeman 
> If I have a route defined in a parent that has chosen to use routing and it has a url such as /mythings/:id then whatever is in :id gets passed into the component's @Input() id:string property.

IMO just using @Input for _all_ things passed in by the router (also RouteData & optional parameters) would be somewhat limiting / dangerous. e.g. if you want to pass some static data to your route you want to be sure that it's not overwritten by a parameter with the same name. Also having to set up an `@Input`-Binding would mean that you need to know about all your parameters at development, which is not necessarily the case. So in the end I personally like having dedicated Classes for `RouteParam` and `RouteData`

On the other hand I totally agree, that reusing a component should be the default case, and using the constructor to pass in these routing-dependent things currently prevents the reuse in most of the cases. Maybe it could be a solution to pass these things in `routerOnActivate` instead of the `constructor`?

**Comment by escardin** [comment]
@choeller I don't think mapping route params to inputs is potentially dangerous, as it stands right now, it isn't possible to bind to those params because the router uses the DCL. They're totally available to be mapped from the route params into those slots. If the component itself changes them, then that's their own problem, just as it would be in a non-routable component.

I personally have always felt that a routable component should protect itself, so as to not require all the routes to that component to implement the same guards. This also ensures that you can't forget to protect that route into the admin section. Others prefer that the route definition has the guards for convenience (or in the case of async components, where the component isn't fetched yet). If components no longer need to be routable, then I think it makes sense that the guard be on the route definition, as the component no longer knows that it's routable (though having a hook on it would be nice anyway).

As to people using `@CanActivate` to block loading of a page until some data has been fetched, I'm not aware of any other way of doing this. It may seem spurious, but it is weird to have a page essentially half navigate because of a slow http request. The more I think about it though, this may be a bad design? I was about to advocate for ngOnInit or similar returning a promise which resolves when it wants to let the component finish drawing.

@CaptainCodeman I was under the impression that if you resolve(/reject?) the promise in `@CanActivate` to a route, it would attempt redirect you there instead.

**Comment by CaptainCodeman** [comment]
> 1. it is already true by default; you don't have to set it true

> That's why I say the only thing you can really do is set it false when you want to ensure that the router does not reuse an instance.

I'm seeing the exact opposite of how you describe it (on beta 12 & since way back alpha4x-ish?). It seems set to false by default and unless I set it to true then the whole control reloads all the time even if the route hasn't changed (but the url has). Basically, I get completely different behaviour if I set it to true vs if I don't set it (so it doesn't seem like true is the default or may be changing something else as well).

I'll try and create a cut down concrete example to demonstrate it but basically I have a route where I load a photo album:

```
{ path: '/folder/*id', component: FolderComponent, name: 'Folder' }
```

All the photos in the album appear as a strip of thumbnails and I don't want those to reload as long as the folder is the same. The selected photo is indicated with a `?photo=x` querystring which updates as a photo is clicked. If canReuse is set to true then just the selected photo updates, if it's set to false or not set then the entire thing updates and all the thumbnails are re-drawn.

**Comment by CaptainCodeman** [comment]
> As to people using @CanActivate to block loading of a page until some data has been fetched, I'm not aware of any other way of doing this. It may seem spurious, but it is weird to have a page essentially half navigate because of a slow http request. The more I think about it though, this may be a bad design? I was about to advocate for ngOnInit or similar returning a promise which resolves when it wants to let the component finish drawing.

Yeah, I think this is one of those things that depends on your app and personal taste. My preference is for something to happen as soon as something has been clicked and if there is some element that can't be rendered because it's still being fetched then it can show a spinner or something but if it just waits on the previous page they start clicking again thinking it's not working (heck, I do that with my own app). I know there are subtle progress loading bars and the like but I don't think people always see them. The pause before something happening makes things "feel" slow IMO.

I can see situations where people want the opposite behaviour though.

**Comment by brandonroberts** [comment]
@CaptainCodeman there are three things that trigger a route component not to be reused. 

1. You navigate to a different route component and come back to the previous route component.
2. You explicitly return/resolve false from `routerCanReuse`.
3. You navigate within the same route component with different parameters.

Link to code: https://github.com/angular/angular/blob/master/modules/angular2/src/router/directives/router_outlet.ts#L156-L172

This would explain why you have to set your `routerCanReuse` to true in order to visit the same route component without reloading. As far as what CanActivate does today, it doesn't do anything if you return/resolve false. It just stops the navigation process for going to the attempted route. The hook for what happens once you fail CanActivate hasn't been implemented yet.

**Comment by CaptainCodeman** [comment]
Thanks @brandonroberts In every case I'm changing the URL because I want it to be bookmarkable / copyable but it's always to the same route as far as the routing should be concerned (i.e. only querystring changes). I'm doubtful what form the reuse takes if you navigate away and back again - it seems like the result would be the same as I'm seeing - everything reloaded and re-rendered (if I go from A to B and back to A, it's no different).

The gaps in functionality such as you describe for CanActivate are part of what makes the router frustrating to use. You can start down a path and then hit a dead end because the road isn't finished. I've seen "we have a plan for this" in issues many times but then months seem to go by without anything materializing.

**Comment by laurelnaiad** [comment]
I don't understand why items 1 and 3 in @brandonroberts list (return visits and changes to parameters) need to be grounds for not reusing components.... it seems to me that if I say my components can be reused, that I'd be trying to assert that items 1 and 3 are taken care of via change detection and other means.  That said, maybe there's a very good reason that I just don't know.

**Comment by brandonroberts** [comment]
@laurelnaiad there isn't a built in caching mechanism for routes. In the case of 1, if you navigate to another route that component is going to be disposed since it was loaded dynamically. So the next time you come back to it, it needs to be re-instantiated. For 3, I think this is a more valid use case. If I'm only changing parameters and nothing else, then it should be reused instead of disposed/re-created.

**Comment by laurelnaiad** [comment]
> there isn't a built in caching mechanism for routes

When i learned of the existence of canReuse, I assumed it was enabling such caching (even in my first comments above).... glad I know now that that's not even its intent. It makes me worry less about setting the flag at all!

**Comment by wardbell** [comment]
Apologies on my statement that `routerCanReuse` defaults to true.  It's more complicated than that.

I [created a plunker](http://plnkr.co/edit/tsqhDjNoD9R9VYonqqTK?p=preview) to explore some of the behavior. Here it is in action.

![can-reuse-anim](https://cloud.githubusercontent.com/assets/129061/14093774/3ca91956-f505-11e5-90d0-2e217c634cb4.gif)

Here's how it works:

There is a shell with three buttons: CanReuse CannotReuse, Default. Clicking them results in a navigation _through a parent component_ to one of three nearly identical children (all derived from the `baseCmp` class). They differ only in `routerCanReuse`:  one implements it returning true, the next implements it returning false, the third doesn't implement it at all (default).

I start with the parent component in the middle that implements
 a `routerCanReuse` that returns false. While in that state, none of its children are ever reused no matter what its `routerCanReuse` method does. 

Then I comment it out  and re-run. Now the router DOES respect a child's `routerCanReuse`.   **This is what made me think that `routerCanReuse` was true by default.**

Clicking the high level button also navigates to an instance of a child, each time with a new id parameter, simulating a request to edit an item from a list. The navigation starts from the root and works down the route tree to the leaf component.

Each of child (leaf) also has a next button that navigates to an instance of its own component time but with a new id parameter. It is as if the component itself could walk the list of items. The point of this feature is to confine the navigation to the leaf component level, unlike the high level buttons which work from the root down. I am trying to discover if there is a difference in behavior between going top down and going sideways.   It turns out that I could not find such a difference.

Only one of the child components is reused ... the `CanReuseCmp`. We can tell because the router calls `ngOnInit` and `routerOnActivate` the first time only; subsequently it calls `routerOnReuse`.

For the other two components, the router always calls `ngOnInit` and `routerOnActivate`; it never calls `routerOnReuse`.  **So for _this_ configuration, the `routerCanReuse` default is false**

Confused? Me too.

**Comment by wardbell** [comment]
After talking to Victor Berchet and stepping through the [pertinent lines](https://github.com/angular/angular/blob/master/modules/angular2/src/router/directives/router_outlet.ts#L156-L172) of the `routerCanReuse` implementation, I understand what is going on and why.

I think the behavior is correct once you know. This has become a documentation issue that I will take on.

**Comment by choeller** [comment]
@escardin I guess I just don't get everything you are saying :grimacing: Can you explain what DCL means? googleing didn't help... 
I would also prefer to use standard mechanism like `@Input` for Binding, but currently I don't see a clean way to support params that are not known at development-time. Also I wonder how to distinguish between `RouteData` and `RouteParams` (that's where I saw a potential problem when one overrides the other). It's quite possible that I just get @CaptainCodeman suggestions wrong, so I'd like to understand more about the approach...

**Comment by petebacondarwin** [comment]
Just to throw a cat among the pigeons I knocked this up... :-)

https://docs.google.com/document/d/1loNVp3tcX_nGb26yz1biSl2z51UcT-kpm5O__tYt0KY/edit

**Comment by escardin** [comment]
@choeller DCL stands for DynamicComponentLoader, which is how the router instantiates and loads components.

Can you point to a use case where you don't know the route params at development time, or where route data and route params have the same values?

**Comment by choeller** [comment]
@escardin Ahhh - ok - thanks :) 

So we have one actual use case in my current project where we don't know about the `RouteParams` at dev-time:
We are builiding a configurable product information sytem as a SAAS where each customer can describe it's datamodel in a DSL and upload it to the server. This config then leads to a custom UI for each customer. e.g. my product has the properties width, height, weight, startOfAvailablityDate etc. We then dynamically create search masks based on the customers datamodel, and when hitting search we create a bookmarkable URL when for displaying the results. This perfectly works with the current architecture where I just take all data from the (one) RouteParams-Object, but I can't figure out how this would look like with `@Input`
Regarding `RouteData` and `RouteParams` I was just thinking about `RouteData` like `needsAdminRights=true` this should not be "overwrittable" by some URL-Query param. Of course one could say "RouteData always wins" but then you would limit the options available for QueryParams (even optional / dynamic ones that you possibly don't know before)
To me the current solution (bundling all RouteParams into one object) perfectly fits the use-cases I see for the router - the only thing I dislike is, that they need to be injected in the constructor...

**Comment by escardin** [comment]
Some of us discussed this in the gitter channel and came to an agreement on how we would like the router to work:
- Move away the concept of routable components and use @Inputs() instead (i.e any component is routable)
- move `@CanActivate()` into the route definition (and optionally add a lifecycle hook in the component like @wardbell  initially described)
- `canReuse` should be deprecated, as the router should just update the inputs instead of routing if the route still matches. (i.e. ngOnChanges and other lifecycle events will be able to handle the change)
- routerOnActivate and routerOnReuse should be deprecated, as when the router navigates away and back, an new instance of the component will be made in that case. (i.e. existing lifecycle events will manage creating and destroying instances)
- CanActivate (wherever it ends up) returns true, a route, or a promise that resolves to true or rejects to a route
- optional route params will just be unset if they are not passed (they could then be set to a default in the constructor). route params that don't map to an input won't be exposed to the component (perhaps an error for mandatory components?)

An alternate proposal is to not use @Inputs() and instead go with a resolve style mechanism, so that the router state is injected as a regular object instead of RouteParams. In this case, navigating to the same route with only different parameters would involve recreating the objects along the chain. This might be resolved by making the input parameters observable.

Overall we want the special router hooks out of the components entirely. Instead we'd prefer to use existing mechanisms to push data into the component, manage it's lifecycle, etc... One exception was made for routerCanActivate as described in the initial issue description as an optional addition to having a canActivate in the route definition. This was because there isn't really an equivalent component construct. On the other hand, having a resolve function in the route definition that could provide injectables to the component would also solve this problem.

**Comment by fxck** [comment]
both router lifecycles and router params as observables would make me happy

**Comment by CaptainCodeman** [comment]
I would go a bit further and suggest that a separate CanActivate and Resolve mechanism aren't necessary and can also be incorporated into the regular component methods if we can make use of parameter inputs suggested. Looking through some of the other issues it turns out this may not be such a crazy idea. Some important issues that it may also help address:

https://github.com/angular/angular/issues/4112#issuecomment-195621363
https://github.com/angular/angular/issues/4112#issuecomment-195780164

The typical use-case for CanActivate usually revolves around some type of per-object permission check that can't be based solely on the route data or sign-in status. So any of the simplistic "does the user need to be authenticated" or "does the user have role xyz" can be statically defined on a route and checked in a router outlet / auth service so doesn't require any of the CanActivate functionality as described here.

When your permission check can't be defined in simple auth status and role terms and instead depends on some data being resolved first things get more complicated. Now we need to use services ("something injected") and have async operations to handle performing the checks. The suggestion is to use CanActivate and augment it with dependency injection mechanisms because the CanActivate method as it stands requires significant workarounds to be made useful. To me this is again adding something that already exists. The component constructor is already configured to have the services that a component requires injected in so I think it makes sense to use it.

The objection to using the component is that the instance needs to be instantiated (unless it's being reused). But I don't think that is as big an issue as it's often portrayed to be. First of all, creating an instance is hardly onerous and most importantly it's usually going to be done anyway if the route navigation succeeds which is the most common outcome by far IMO. The services need to be injected, the item needs to be requested. All the things a smart component was going to do anyway. Why not just let it do it? The DI / resolves of a CanActivate method simply delays doing this and it seems to me becomes focused, I think mistakenly, in optimizing the "failure" path rather than the "success" path.

Think about why you want to reject a route navigation: a user doesn't have permission to the requested item. Why are they navigating to the route that shouldn't be navigated to? Maybe we listed an item they don't have permission to see (UI failure elsewhere), or they are hacking the URL to get access to something they shouldn't or have obtained a link to something they don't have access too. In all cases, why are we optimizing this failure path? Why do we care if things take a little longer for them and "waste" instantiating a component for a request that is going to fail? Why not just let the normal operation proceed up to the point - component created & services injected (or component reused), input parameters set and ngOnChanges called. At this point the smart component does it's normal stuff and loads the data but can then decide to carry on or reject the request based on the result it gets. Maybe by returning something from ngOnChanges (which could be a promise to delay the route loading & resolve to false to signify route rejection). Can that satisfy the resolve delay and route rejection?

If all is well though (the typical case) then the request proceeds normally and the component is loaded and rendered with the simplest code-path possible. I honestly can't come up with good legitimate example where rejecting routes is the normal and common use-case that makes doing it all as a separate pre-check simply to avoid creating a component instance worth the complexity that the separate method introduces.

I think that in the vast majority of cases when the normal course of operation is that a route navigation proceeds this is the simplest and most straightforward way. All the code to do with the smart component is kept within the component as are service dependencies and the DI level of them. Nothing is spread out to route configs. The component rules and the security doesn't depend on the router being the thing to instantiate the component - it works in all cases, even if the component is moved and hosted by a regular parent component instead.

I sense this is more controversial suggestion. I think it involves letting go of some notion that we must have this because we've always had it before. I'm really just talking about moving where it's done into the component instance and not making it a separate method - just allow the existing method to pass some information back to signify a delay and rejection if required.

**Comment by cquillen2003** [comment]
@wardbell IMHO the following logic cited above neglects the other side of the coin:

"You can't use a router independently of a component so let's make it easy and do both router configuration and component definition in the same place."

True.  But you can use a component without a router.  I think this is a better way to look at it because components are the fundamental building block.

**Comment by wardbell** [comment]
@CaptainCodeman - You make a persuasive argument for evaluating the *can activate* logic within the target component. Those on the other side of that argument prefer to block activation earlier, especially because the route may involve async loading which they feel is too heavy a price to pay for a failed route even if failed routes are rare.  FWIW, the issue has been decided and we do well to move on.

@cquillen2003 - I don't follow you. I was reflecting the argument in favor of writing router configuration within a component rather than in some other place. If you're writing "a component without a router" it's all moot, right? There's nothing to configure and nothing to put in (or keep out) of the component. 

@escardin As attractive as using inputs sounds, it doesn't hold up in many common use cases. It's not clear how route or query string parameters should align with input properties nor how the URL material that doesn't match input properties should become available to the component. I fear for the complexity of whatever might be proposed for matching input properties to parameters buried in a URL. 

You may have neglected the critical role of the query string in a URL. These shouldn't be ignored nor can they be parsed into input properties. I believe this is the essence of @choeller's critique of your proposal.

Meanwhile, all of the tried-and-true routers in my experience keep the routing constructs separate from the component/controller.  We know this separation works.

I think you misunderstand the role of `canReuse`.  It participates in the router's decision about whether a component instance can be recycled or must be discarded and recreated. There is no equivalent among the component lifecycle hooks.

The component lifecycle hooks are no substitute for `routerOnActivate` and `routerOnReuse`. They provide valuable information about the navigation to and from the target component that is simply not available any other way, neither through the component hooks nor through input params.

To be perfectly honest, worthy as some of these ideas may be, they are too nebulous and incomplete.  They are very far from addressing concretely the routing scenarios we know we need to handle.

You may disagree. That's cool. I don't wish to be cruel and hard-hearted but ... these ideas arrive too late. At this point in the game, the design of the router has moved on and I think I can say definitively that

* it won't be template-driven; the @petebacondarwin cat is dead (apologies to Schrödinger)

* `@CanActivate` will become something like a `canActivateChildRoutes` instance method of a routing component.

* the routing hooks `routerOnActivate` and `routerOnReuse` will live on

* routing parameters will live on (although they don't have to be injected as they are available in the instruction parameters of the router lifecycle hooks).

* input properties will remain independent of the routing structure

* components that need to know information derived from navigation will have to become aware of the routing subsystem.

It's time to move on.

It may also be worth remembering that the Component Router is an optional subsystem. It's not part of the Angular core. If it sucks, Angular will survive, someone will build a better one, and we'll all move to that.

My personal bet is on the Component Router which I feel soon will overcome it's current, most pressing defects.

**Comment by CaptainCodeman** [comment]
I guess the Component Router *didn't* overcome it's defects as it's now deprecated. Given that the new replacement router also changes how lazy-loading is going to happen and the services that may be needed to make the decisions could be loaded with it (?) then it seems to me that the only place the `CanActivate` type decision can *consistently* be supported is on the component itself.

So, I'd go back to my suggestion of optimizing for the most typical, successful code path (instead of optimizing for failure) by simply allowing a promise to be returned to trigger rejection and / or some event raised through the router outlet that the parent component can use.

**Comment by vsavkin** [comment]
We are no longer maintaining the deprecated router to focus our efforts on making the new router solid.

Activation is done via guards in the new router.

Closing this issue.
